beacon-chain/p2p/discovery.go

```diff
@@ -330,72 +324,127 @@ func (s *Service) listenForNewNodes() {
+// peers for each subnet listed in `subnets`.
+// If, for all subnets, the threshold is met, then this function immediately returns.
+// Otherwise, it searches for new peers for defective subnets, and dials them.
+func (s *Service) findPeers(ctx context.Context, targetPeerCount uint) error {
+	// Return early if the discovery listener isn't set.
+	if s.dv5Listener == nil {
+		return nil
+	}
+
+	// Restrict dials if limit is applied.
+	maxConcurrentDials := math.MaxInt
+	if flags.MaxDialIsActive() {
+		maxConcurrentDials = flags.Get().MaxConcurrentDials
+	}
+
+	// Compute the number of new peers we want to dial.
+	missingPeerCount := s.missingPeerCount(targetPeerCount)
+	for missingPeerCount > 0 {
+		if err := func() error {
+			// Create an discovery iterator to find new peers.
+			iterator := s.dv5Listener.RandomNodes()
+
+			// The peer crawling stops when enough peers are found in all defective subnets,
+			// or when `batchPeriod` is reached.
+			// It is useful to dial peers even if not all peers in defective subnets
+			// are found.
+			ctx, cancel := context.WithTimeout(ctx, batchPeriod)
+			defer cancel()
+
+			// `iterator.Next` can block indefinitely. `iterator.Close` unblocks it.
+			// So it is important to close the iterator when the context is done to ensure
+			// that the search does not hang indefinitely.
+			go func() {
+				<-ctx.Done()
+				iterator.Close()
+			}()
+
+			// Crawl the network for peers subscribed to the defective subnets.
+			nodeByNodeID := make(map[enode.ID]*enode.Node)
+			for missingPeerCount > 0 && iterator.Next() {
+				if ctx.Err() != nil {
+					// If the context is done, we stop iterating.
+					break
+				}
+				// Iterate on peers and skip those that do not match the filter.
+				node := iterator.Node()
+				if !s.filterPeer(node) {
 					continue
 				}
 
-				if peerInfo == nil {
+				// Remove duplicates, keeping the node with higher seq.
+				existing, ok := nodeByNodeID[node.ID()]
+				if ok && existing.Seq() > node.Seq() {
 					continue
 				}
+				nodeByNodeID[node.ID()] = node
+
+				// We found a new peer. Modify the defective subnets map
+				// and the filter accordingly.
+				missingPeerCount--
+			}
 
-				// Make sure that peer is not dialed too often, for each connection attempt there's a backoff period.
-				s.Peers().RandomizeBackOff(peerInfo.ID)
-				wg.Add(1)
-				go func(info *peer.AddrInfo) {
-					if err := s.connectWithPeer(s.ctx, *info); err != nil {
-						log.WithError(err).Tracef("Could not connect with peer %s", info.String())
-					}
-					wg.Done()
-				}(peerInfo)
+			// Convert the map to a slice.
+			peersToDial := make([]*enode.Node, 0, len(nodeByNodeID))
+			for _, node := range nodeByNodeID {
+				peersToDial = append(peersToDial, node)
 			}
-			wg.Wait()
+
+			// Dial new peers in batches.
+			s.dialPeers(maxConcurrentDials, peersToDial)
+
+			return nil
+		}(); err != nil {
+			return err
+		}
+
+		if err := ctx.Err(); err != nil {
+			return err
```

Comment:
this is when the context cancels right? should we wrap or replace with our own message if we know why

Reply:
Fixed in Fixed in ee3857631d4403b42cd3348846e96784d0bdde0d.
The only reason for the error to be not nil is parent context cancellation.
(Added some info in the godoc for the caller.)